src/arch/i386/image/freebsd.c

```diff
@@ -0,0 +1,203 @@
+ */
+
+FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );
+
+#include <errno.h>
+#include <elf.h>
+#include <stdio.h>
+#include <ipxe/image.h>
+#include <ipxe/elf.h>
+#include <ipxe/features.h>
+#include <ipxe/init.h>
+#include <ipxe/io.h>
+
+typedef uint64_t p4_entry_t;
+typedef uint64_t p3_entry_t;
+typedef uint64_t p2_entry_t;
+extern p4_entry_t PT4[];
+extern p3_entry_t PT3[];
+extern p2_entry_t PT2[];
+extern char * BLAH;
+
+uint32_t entry_hi;
+uint32_t entry_lo;
+
+extern void amd64_tramp();
+#define	VTOP(va)	virt_to_phys(va)
+typedef char * caddr_t;
+void __exec(caddr_t, ...);
+
+#define ALIGN_UP(value, align)  (((value) & (align-1)) ?                \
+                                (((value) + (align-1)) & ~(align-1)) : \
+                                (value))
+#define ALIGN_PAGE(a)           ALIGN_UP (a, 4096)
+
+#define PG_V	0x001
+#define PG_RW	0x002
+#define PG_U	0x004
+#define PG_PS	0x080
+
+/**
+ * @file
+ *
+ * ELF bootable image
+ *
+ */
+
+FEATURE ( FEATURE_IMAGE, "ELF64", DHCP_EB_FEATURE_ELF, 1 );
+
+/**
+ * Prepare segment for loading
+ *
+ * @v segment		Segment start
+ * @v filesz		Size of the "allocated bytes" portion of the segment
+ * @v memsz		Size of the segment
+ * @ret rc		Return status code
+ */
+static int allocate_segment ( physaddr_t * segment, size_t sz ) {
+	struct memory_map memmap;
+	unsigned int i;
+
+	get_memmap ( &memmap );
+
+	/* Look for a suitable memory region */
+	for ( i = 0 ; i < memmap.count ; i++ ) {
+		if ( (memmap.regions[i].end - memmap.regions[i].start) > sz) {
+			*segment = memmap.regions[i].start;
+			memset_user ( *segment, 0, 0, sz);
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+/**
+ * Execute ELF image
+ *
+ * @v image		ELF image
+ * @ret rc		Return status code
+ */
+static int elfboot64_exec ( struct image *image ) {
+	int i;
+
+	size_t size = 0;
+	physaddr_t buffer = 0;
+
+	unsigned long kern_end, modulep = 0;
+	struct image *module_image;
+	Elf64_Ehdr ehdr;
+
+	// Get total size
+	for_each_image ( module_image ) {
+		size += ALIGN_PAGE(image->len);
+	}
+	//XXX: size += ALIGN_PAGE(params)
+	//XXX: size += ALIGN_PAGE(env)
+
+	if (allocate_segment(&buffer, size) != 0) {
+		DBG ( "Couldn't allocate enough memory to fit kernel, needed %d bytes", size);
+		return -1;
+	}
+
+
+	modulep = buffer;
+
+	// Copy kernel
+	memcpy_user(phys_to_user(buffer), 0, image->data, 0, image->len);
+	buffer += ALIGN_PAGE(image->len);
+	copy_from_user ( &ehdr, image->data, 0, sizeof ( ehdr ) );
+
+	// Copy image
+	for_each_image ( module_image ) {
+		if (image == module_image) continue;
+		memcpy_user(phys_to_user(buffer), 0, module_image->data, 0, module_image->len);
+		buffer += ALIGN_PAGE(module_image->len);
+	}
+
+	kern_end = buffer;
+
+	entry_lo = ehdr.e_entry & 0xffffffff;
+	entry_hi = (ehdr.e_entry >> 32) & 0xffffffff;
+
+	for (i = 0; i < 512; i++) {
+		/* Each slot of the level 4 pages points to the same level 3 page */
+		PT4[i] = (p4_entry_t)VTOP((void *)&PT3[0]);
```

Comment:
For this one, I need to get back the physical address of PT3, see https://svnweb.freebsd.org/base/head/sys/boot/i386/libi386/elf64_freebsd.c?view=markup#l98.

vtop is defined https://svnweb.freebsd.org/base/head/sys/boot/i386/btx/lib/btxv86.h?view=markup#l57 and __base https://svnweb.freebsd.org/base/head/sys/boot/i386/btx/lib/btxcsu.S?view=markup#l39

It looks like virtual_to_phys does not behave the same.